org.eclipse.xtext.tests/src/org/eclipse/xtext/nodemodel/serialization/ResourceStorageTest.java

```diff
@@ -0,0 +1,185 @@
+import org.eclipse.xtext.testlanguages.SimpleExpressionsTestLanguageStandaloneSetup;
+import org.eclipse.xtext.testlanguages.simpleExpressions.Expression;
+import org.eclipse.xtext.tests.AbstractXtextTests;
+import org.eclipse.xtext.util.StringInputStream;
+import org.junit.Test;
+
+import com.google.inject.Guice;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+
+public class ResourceStorageTest extends AbstractXtextTests {
+	
+	public static class CustomSetup extends SimpleExpressionsTestLanguageStandaloneSetup {
+		@SuppressWarnings("unused")
+		@Override
+		public Injector createInjector() {
+			return Guice.createInjector(new SimpleExpressionsTestLanguageRuntimeModule() {
+				public Class<? extends IResourceStorageFacade> bindIResourceStorageFacade() {
+					return ResourceStorageFacade.class;
+				}
+				@Override
+				public Class<? extends XtextResource> bindXtextResource() {
+					return StorageAwareResource.class;
+				}
+				public Class<? extends AbstractFileSystemAccess2> bindAbstractFileSystemAccess2() {
+					return JavaIoFileSystemAccess.class;
+				}
+				public Class<? extends ParseResultWrapper> bindParseResultWrapper() {
+					return StorageAwareDetachableParseResultWrapper.class;
+				}
+				public Class<? extends NodeModelBuilder> bindNodeModelBuilder() {
+					return DetachableNodeModelBuilder.class;
+				}
+				public Class<? extends PartialParsingHelper> bindPartialParsingHelper() {
+					return null;
+				}
+			});
+		}
+	}
+	
+
+	public static class InMemoryURIConverter extends ExtensibleURIConverterImpl {
+		private final Map<URI, InputStream> models = new HashMap<URI, InputStream>();
+
+		public void addModel(String uri, String content) {
+			models.put(URI.createURI(uri), new StringInputStream(content));
+		}
+
+		@Override
+		public boolean exists(URI uri, Map<?, ?> options) {
+			boolean result = models.containsKey(uri);
+			if (!result) {
+				for (URI knownUri : models.keySet()) {
+					if (uri.toString().endsWith(knownUri.toString())) {
+						return true;
+					}
+				}
+			}
+			return result;
+		}
+
+		@Override
+		public InputStream createInputStream(URI uri, Map<?, ?> options) throws IOException {
+			return models.get(uri);
+		}
+	}
+	
+	private ParserTestHelper helper;
+
+	@Inject
+	private ResourceStorageFacade resourceStorageFacade;
+	
+	@Override
+	public void setUp() throws Exception {
+		super.setUp();
+		with(CustomSetup.class);
+		helper = new ParserTestHelper(getResourceFactory(), getParser(), get(Keys.RESOURCE_SET_KEY),getCurrentFileExtension());
+		injectMembers(this);
+	}
+
+	@Test
+	public void testWriteAndLoad_01() throws Exception {
+		String modelAsString = "(d - e) / e * d // fasdf s";
+		doTestWriteAndLoad(modelAsString);
+	}
+	
+	@Test
+	public void testWriteAndLoad_02() throws Exception {
+		String modelAsString = "a + b + c + d + e +";
+		doTestWriteAndLoad(modelAsString);
+	}
+	
+	void doTestWriteAndLoad(String modelAsString) throws Exception {
```

Comment:
I opted for duplicating some code in the test cases to see proper. I'm not sure it would be worthwhile to extract parts since the tests wouldn't be self-contained anymore.

Reply:
As I always say, but I'm not the author of that, "code duplication is admissible in tests if that makes them more readable." :)